### PR TITLE
FD-74: Add channels(ingest and fetch)  and rules to client.

### PR DESCRIPTION
**Pull request by ian-sift**
Expecting Ailin to review Channels and Ingestion files and Marc to review Rules. (of course welcome to cross review)

#### Channel Fetch
```
Getting data for multiple channels:
0: voltage: 100 points. Avg: 5.99
1: mainmotor.velocity: 29 points. Avg: 5.586206896551724
Start time: 2025-07-02 21:17:28.878926+00:00, End time: 2025-07-02 21:18:24.878785+00:00
2: gpio.12v: 89 points. Avg: 1.0
3: gpio.charge: 178 points. Avg: 0.8932584269662921
4: gpio.heater: 267 points. Avg: 0.6666666666666666
5: gpio.led: 356 points. Avg: 1.3089887640449438

=== Testing cache with varying time ranges ===

Test 1: Exact same time range no run_id (should hit cache)

Test 2: Subset of time range (should hit cache if overlapping)

Test 3: Extended time range earlier (should hit cache for overlapping portion)

Test 4: Different time encompassed range (should hit cache)

Test 5: No time range specified (should miss cache)
0: gpio.12v: 100 points. Avg: 1.0
1: gpio.charge: 100 points. Avg: 0.88
2: gpio.heater: 100 points. Avg: 0.19
3: gpio.led: 100 points. Avg: 3.07
4: voltage: 100 points. Avg: 5.02
5: mainmotor.velocity: 100 points. Avg: 5.24

Test 6: No time range specified again (should hit cache)
0: gpio.12v: 189 points. Avg: 1.0
1: gpio.charge: 189 points. Avg: 0.8359788359788359
2: gpio.heater: 189 points. Avg: 0.20105820105820105
3: gpio.led: 189 points. Avg: 3.1481481481481484
4: voltage: 200 points. Avg: 5.505
5: mainmotor.velocity: 129 points. Avg: 5.317829457364341

=== Cache Performance Summary ===
Original call: 0.3581 seconds
Exact time range no run_id: 0.0039 seconds (92.2x faster)
Subset time range: 0.0033 seconds (107.2x faster)
Extended time range earlier: 0.0892 seconds (4.0x faster)
Different time range: 0.0049 seconds (73.5x faster)
No time range: 0.8140 seconds (2.3x slower)
No time range repeat: 0.0023 seconds (360.6x faster) 
```

#### Channel Ingest
```
2025-07-22 18:53:27,727 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Getting ingestion config id from generated client key 8d85cbd7c4f302bb962bfa5f9b22d8c8658c1a6c7c9fb989960e789b655d6113
2025-07-22 18:53:27,917 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion config cb2850f1-6d84-45b0-a2f7-114996992412 already exists but is not yet in the cache.
2025-07-22 18:53:28,473 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Built new stream for ingestion config cb2850f1-6d84-45b0-a2f7-114996992412
Flow test-flow ingestion config id: cb2850f1-6d84-45b0-a2f7-114996992412, asset: ian-test-asset
Flow highspeed-flow ingestion config id: cb2850f1-6d84-45b0-a2f7-114996992412, asset: ian-test-asset
2025-07-22 18:53:28,474 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread started
2025-07-22 18:53:28,475 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Getting ingestion config id from generated client key 3285fd28e95919241e9a717f2e828a402461aed33be9306c5b8a201ec2cca4ac
2025-07-22 18:53:28,578 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion config 68f27ca2-c90c-4624-bb25-44c4d78f194a already exists but is not yet in the cache.
2025-07-22 18:53:29,183 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Built new stream for ingestion config 68f27ca2-c90c-4624-bb25-44c4d78f194a
Flow new-asset-flow ingestion config id: 68f27ca2-c90c-4624-bb25-44c4d78f194a, asset: test-asset-ian2
2025-07-22 18:53:29,201 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread started
2025-07-22 18:53:29,258 - sift_client._internal.low_level_wrappers.ingestion - INFO - Ingestion thread finishing.
2025-07-22 18:53:29,263 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 100 requests, remaining: 2452
2025-07-22 18:53:29,267 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 200 requests, remaining: 2352
2025-07-22 18:53:29,272 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 300 requests, remaining: 2252
2025-07-22 18:53:29,276 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 400 requests, remaining: 2152
2025-07-22 18:53:29,281 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 500 requests, remaining: 2052
2025-07-22 18:53:29,285 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 600 requests, remaining: 1952
2025-07-22 18:53:29,289 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 700 requests, remaining: 1852
2025-07-22 18:53:29,293 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 800 requests, remaining: 1752
2025-07-22 18:53:29,298 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 900 requests, remaining: 1652
2025-07-22 18:53:29,302 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1000 requests, remaining: 1552
2025-07-22 18:53:29,306 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1100 requests, remaining: 1452
2025-07-22 18:53:29,311 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1200 requests, remaining: 1352
2025-07-22 18:53:29,315 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1300 requests, remaining: 1252
2025-07-22 18:53:29,319 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1400 requests, remaining: 1152
2025-07-22 18:53:29,323 - sift_client._internal.low_level_wrappers.ingestion - DEBUG - Ingestion thread sent 1500 requests, remaining: 1052
Ingestion time: 0:00:00.123401 seconds
Ingested 2650 datapoints
Ingestion rate: 21471.92 datapoints/second
```

#### Rules
```

=== Creating 8 rules with unique suffix: 20250722185633 ===
Created rule: test_rule_20250722185633_0 (ID: a502743f-1374-4227-a332-dfc780dfee7f)
Created rule: test_rule_20250722185633_1 (ID: 394745ea-cb4f-4ab3-bec5-d3477f663fd7)
Created rule: test_rule_20250722185633_2 (ID: 65a7f5db-0781-4e01-b4b6-f2ca989ae317)
Created rule: test_rule_20250722185633_3 (ID: 06ef7328-cfc7-42e8-9a48-51a8e80db4f1)
Created rule: test_rule_20250722185633_4 (ID: e0eaa9e3-51ce-49ca-8ef4-cbdc0cab91da)
Created rule: test_rule_20250722185633_5 (ID: 8d58e711-8d53-4551-9e87-fabb02fee23d)
Created rule: test_rule_20250722185633_6 (ID: 0e4fb131-9512-48bd-a400-ebd651af63dc)
Created rule: test_rule_20250722185633_7 (ID: f25e02ad-1121-4296-b850-927340a8f666)

=== Testing comprehensive update scenarios ===

--- Test 1: Update expression and channel references ---
Updated : expression = $1 > 0.5

--- Test 2: Update description ---
Updated : description = Updated description with more details about velocity-to-voltage ratio monitoring

--- Test 3: Update action ---
Updated : action type = RuleActionType.ANNOTATION
  - annotation type: RuleAnnotationType.PHASE
  - tags: ['updated', 'phase', 'alert']
  - assignee: 1eba461b-fa36-4e98-8fe8-ff32d3e43a6e

--- Test 4: Update name ---
Updated renamed_rule_20250722185633_4 -> renamed_rule_20250722185633_4

--- Test 5: Update multiple fields simultaneously ---
Updated :
  - description: Multi-field update test
  - is_enabled: True
  - version_notes: 

--- Test 6: Update with complex expression ---
Updated : complex expression = $1 > 0.3 && $1 < 0.8

--- Test 7: Update action to notification ---

--- Test 8: Update tag_ids and contextual_channels ---
Updated :
  - tag_ids: []
  - contextual_channels: ['temperature', 'pressure']

--- Test 8b: Edge case - Invalid expression test ---
Invalid expression update failed as expected: <AioRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "invalid argument: No valid inputs exist for expression: invalid_expression (febc8278-a85f-42b9-9eff-c279e979d368)"
        debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"invalid argument: No valid inputs exist for expression: invalid_expression (febc8278-a85f-42b9-9eff-c279e979d368)", grpc_status:3}"
>

--- Test 9: Batch operations demonstration ---
Batch retrieved 8 rules:
  - : $1 > 0.5
  - : $1 > 0.1
  - : $1 > 0.1
  - renamed_rule_20250722185633_4: $1 > 0.1
  - : $1 > 0.1
  - : $1 > 0.3 && $1 < 0.8
  - test_rule_20250722185633_6: $1 > 0.1
  - : $1 > 0.1

--- Test 10: Delete rules ---

=== Test Summary ===
Created: 8 rules
Updated: 8 rules

=== Validation Checks ===
Traceback (most recent call last):
  File "/Users/ian/code/sift/python/lib/sift_client/tests/integrated/rules.py", line 277, in <module>
    main()
    ~~~~^^
  File "/Users/ian/code/sift/python/lib/sift_client/tests/integrated/rules.py", line 260, in main
    assert updated_rule_5.is_enabled == False, (
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Enabled state update failed: True
```

### CODE REVIEW

**Comment by ian-sift** [comment]
TODO: Rule disabling not working

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+            page_size = max_results
+        return await self._handle_pagination(
+            self.list_channels,
+            kwargs={"query_filter": query_filter},
+            page_size=page_size,
+            order_by=order_by,
+            max_results=max_results,
+        )
+
+    # TODO: Cache calls. Only read cache if end_time is more than 30 min in the past.
```
Comment (niliayu) [inline python/lib/sift_client/_internal/low_level_wrappers/channels.py:177]: When are we thinking of addressing this TODO? Just want to be sure we're able to follow up on these when we intend to since things like this will be quite nice.

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+                    # Case 3
+                    return (None, start_time, end_time)
+            else:
+                if start_time_cached < end_time and end_time <= end_time_cached:
+                    # Case 4
+                    ret_start_time = start_time
+                    ret_end_time = start_time_cached
+                else:
+                    # Case 5
+                    return (None, start_time, end_time)
```
Comment (niliayu) [inline python/lib/sift_client/_internal/low_level_wrappers/channels.py:277]: Maybe a future optimization could be, for the second example in the drawing for Case 5, to just query for the bits outside the beginning/end of the cached window? Luck implemented this behavior in the Grafana plugin so you could get his advice on this.

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        self._stop = threading.Event()
+        self.sift_stream_builder = sift_stream_builder
+        self.ingestion_config = ingestion_config
+        self.no_data_timeout = timedelta(seconds=no_data_timeout)
+        self.metric_interval = timedelta(seconds=metric_interval)
+
+    def stop(self):
+        self._stop.set()
+        # Give a brief chance to finish the stream (should take < 50ms).
+        time.sleep(0.16)
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:94]: it envelopes the no-data sleep time + avg time of stream.finish(). I can break this into constants

File: python/lib/sift_client/examples/generic_workflow_example.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+from sift_client.types import (
+    CalculatedChannelUpdate,
+    ChannelReference,
+    RuleAction,
+    RuleAnnotationType,
+    RuleUpdate,
+)
+
+"""
+Placeholder for future examples.
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/examples/generic_workflow_example.py:17]: ya. Internal only until. Sort of deferring till FD-67

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+        """
+        Get the ingestion config id.
+        """
+        filter_query = cel.equals("client_key", client_key)
+        ingestion_configs = await self.list_ingestion_configs(filter_query)
+        if not ingestion_configs:
+            return None
+        if len(ingestion_configs) > 1:
+            raise ValueError(
+                f"Expected 1 ingestion config for client key {client_key}, got {len(ingestion_configs)}"
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:245]: I can still say reach out to sift team. I don't think `client_key ==` should ever have > 1 returns

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py (hunk 1 above)
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/channels.py:177]: Will remove this todo in favor of an entry about caching on our internal notes page

File: python/lib/sift_client/_internal/low_level_wrappers/rules.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+        for rule_update in rules:
+            rule = await self.get_rule(rule_id=rule_update.resource_id)
+            request = self._update_rule_request_from_update(rule, rule_update)
+            update_requests.append(request)
+
+        request = BatchUpdateRulesRequest(rules=update_requests)  # type: ignore
+        response = await self._grpc_client.get_stub(RuleServiceStub).BatchUpdateRules(request)
+        return cast(BatchUpdateRulesResponse, response)
+
+    async def delete_rule(self, rule_id: str | None = None, client_key: str | None = None) -> None:
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/rules.py:304]: rename to archive (which is the actual action performed)

File: python/lib/sift_client/_internal/low_level_wrappers/rules.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+        request_kwargs: dict[str, Any] = {}
+        if rule_ids is not None:
+            request_kwargs["rule_ids"] = rule_ids
+        if client_keys is not None:
+            request_kwargs["client_keys"] = client_keys
+
+        request = BatchDeleteRulesRequest(**request_kwargs)
+        await self._grpc_client.get_stub(RuleServiceStub).BatchDeleteRules(request)
+
+    async def undelete_rule(
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/rules.py:352]: "unarchive"

File: python/lib/sift_client/types/rule.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+    # Fields from proto
+    rule_id: str | None = None
+    created_date: datetime | None = None
+    modified_date: datetime | None = None
+    created_by_user_id: str | None = None
+    modified_by_user_id: str | None = None
+    organization_id: str | None = None
+    rule_version: RuleVersion | None = None
+    deleted_date: datetime | None = None
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:59]: rename this to "archived_date" since it is misleading

File: python/lib/sift_client/types/rule.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    is_external: bool | None = None
+
+    @property
+    def is_deleted(self) -> bool:
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:63]: "is_archived"

File: python/lib/sift_client/types/rule.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        Args:
+            update: Either a RuleUpdate instance or a dictionary of key-value pairs to update.
+        """
+        updated_rule = self.client.rules.update(
+            rule=self, update=update, version_notes=version_notes
+        )
+        self._update(updated_rule)
+        return self
+
+    def delete(self) -> None:
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:80]: "archive"

File: python/lib/sift_client/types/rule.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+)
+
+from sift_client.types._base import BaseType, ModelUpdate
+from sift_client.types.channel import ChannelReference
+
+if TYPE_CHECKING:
+    from sift_client.client import SiftClient
+
+
+class Rule(BaseType[RuleProto, "Rule"]):
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:35]: For consideration, add properties such as:
- organization
- created_by
- modified_by
- tags
- assets

File: python/lib/sift_client/types/rule.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    """
+
+    name: str
+    description: str
+    is_enabled: bool = True
+    expression: str | None = None
+    channel_references: List[ChannelReference] | None = None
+    action: RuleAction | None = None
+    asset_ids: List[str] | None = None
+    tag_ids: List[str] | None = None
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:47]: rename to `asset_tag_ids` since this is not tags on the rule, but on the asset scoping. It may make sense to have a nested set of fields around asset scoping similar to the proto.

File: python/lib/sift_client/types/rule.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    created_by_user_id: str | None = None
+    modified_by_user_id: str | None = None
+    version_id: str | None = None
+    annotation_type: RuleAnnotationType | None = None
+    tags: List[str] | None = None
+    assignee: str | None = None
+
+    @classmethod
+    def annotation(
+        cls, annotation_type: RuleAnnotationType, tags: List[str] = [], assignee: str | None = None
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:202]: rename to "annotation_default_assignee_user_id" for clarity

File: python/lib/sift_client/types/rule.py [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+            ),
+            action_type=action_type,
+            annotation_type=RuleAnnotationType.from_str(
+                proto.configuration.annotation.annotation_type  # type: ignore
+            )
+            if action_type == RuleActionType.ANNOTATION
+            else None,
+        )
+
+    def to_update_proto(self) -> UpdateActionRequest:
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:246]: Private method - should this be a superclass abstract method?

File: python/lib/sift_client/types/rule.py [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+        return cls(int(val))
+
+
+class RuleAction(BaseModel):
+    """
+    Model of a Rule Action.
+    """
+
+    model_config = ConfigDict(frozen=True, arbitrary_types_allowed=True)
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:187]: Why is abitrary_types_allowed set to True?

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+            ingestion_config: The ingestion config to use for ingestion.
+        """
+        data_queue: Queue[List[IngestWithConfigDataStreamRequestPy]] = Queue()
+        existing = self.stream_cache.get(ingestion_config_id)
+        if existing:
+            existing_data_queue, existing_ingestion_config, existing_thread = existing
+            if existing_thread.is_alive():
+                return existing_thread
+            else:
+                ingestion_config = existing_ingestion_config
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:272]: left as an assert since a user should never really see this vs something they could see (per a pr comment form Marc elsewhere).

File: python/lib/sift_client/types/rule.py (hunk 15 above)
Comment (ian-sift) [reply python/lib/sift_client/types/rule.py:198]: undoing

File: python/lib/sift_client/resources/ingestion.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            flow._apply_client_to_instance(self.client)
+            if run_id:
+                flow.run_id = run_id
+
+        return ingestion_config_id
+
+    def ingest(
+        self,
+        *,
+        flow: Flow | None = None,
```
Comment (marc-sift) [inline python/lib/sift_client/resources/ingestion.py:85]: I don't think it makes sense to have multiple flows specified here. channel_values should map to just a single flow. Overall, I think it might make sense to create separate a `FlowConfig` and ` Flow` class. The former is used to register your ingestion config, the latter is used to create actual values to be ingested. Then you could pass in multiple Flows to ingest at once.

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py (hunk 2 above)
Comment (niliayu) [reply python/lib/sift_client/_internal/low_level_wrappers/channels.py]: Yeah, I think it'll depend on the situation a bit -- if a user is only requesting a small time window it might be the case that one contiguous call is better, but if they're requesting large chunks of data over large time spans, 2 separate smaller calls may be the better option. maybe something we can wait and see how customers are using this API most often to see what the best approach will be.

File: python/lib/sift_client/tests/integrated/ingestion.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+async def main():
+    client = SiftClient(
+        grpc_url=os.getenv("SIFT_GRPC_URI", "localhost:50051"),
+        api_key=os.getenv("SIFT_API_KEY", ""),
+        rest_url=os.getenv("SIFT_REST_URI", "localhost:8080"),
+    )
+
+    asset = "ian-test-asset"
+
+    # TODO:Get user id from current user
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/tests/integrated/ingestion.py:26]: Probably, but also a service user has other benefits in terms of making it easier to find all side effects of a test. Wanna defer this work to FD-59

File: python/lib/sift_client/resources/ingestion.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        """
+        Initialize the IngestionAPI.
+
+        Args:
+            sift_client: The Sift client to use.
+        """
+        super().__init__(sift_client)
+        self._low_level_client = IngestionLowLevelClient(grpc_client=self.client.grpc_client)
+
+    # TODO: How to flag this can't be called from the sync API?
```
[the remark this answers is not in the document]
Comment (ian-sift) [reply python/lib/sift_client/resources/ingestion.py:38]: I'll use the async. Thoughts on including async in the top level client so you can do sift.client.ingestion vs sift.client.async_.ingestion?

[the remark this answers is not in the document]
Comment (niliayu) [reply python/lib/sift_client/resources/ingestion.py]: I'm for it!

**Review by niliayu — COMMENTED** [review]
Looking really good! One nit left on the deps

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+        Args:
+            channel_id: The channel ID to get.
+
+        Returns:
+            The Channel.
+
+        Raises:
+            ValueError: If channel_id is not provided.
+        """
+        if not channel_id:
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/channels.py:54]: These kind of checks will get onerous and we can leverage what we get for free with 1) Python argument count check, 2) API error on empty string.

Also no need to do this at the low-level client since we are just wrapping the API so we should leverage the APIs/proto generated code behavior for validation, etc.

File: python/lib/sift_client/_internal/low_level_wrappers/channels.py [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+        channels = [Channel._from_proto(channel) for channel in response.channels]
+        self._update_name_id_map(channels)
+        return channels, response.next_page_token
+
+    async def list_all_channels(
+        self,
+        *,
+        query_filter: str | None = None,
+        order_by: str | None = None,
+        max_results: int | None = None,
```
[the remark this answers is not in the document]
Comment (alexluck-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/channels.py:106]: I was just trying to keep this low-level as a wrapper and not make any assumptions, but no strong feelings here. This is internal so we can always update page size in the future

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+        self.sift_stream_builder = sift_stream_builder
+        self.ingestion_config = ingestion_config
+        self.no_data_timeout = no_data_timeout
+        self.metric_interval = metric_interval
+        self.initialized = False
+
+    def stop(self):
+        self._stop_event.set()
+        # Give a brief chance to finish the stream (should take < 50ms).
+        time.sleep(self.CLEANUP_TIMEOUT)
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:86]: should this be asyncio.sleep?

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+    async def await_stream_build(self):
+        while not self.initialized:
+            await asyncio.sleep(0.01)
+
+    async def main(self):
+        logger.debug("Ingestion thread started")
+        self.sift_stream_builder.ingestion_config = self.ingestion_config
+        sift_stream = await self.sift_stream_builder.build()
+        time_since_last_metric = time.time() - 1
+        time_since_last_data = time.time()
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:98]: isn't this an absolute value and not a delta? Shouldn't it be zero?

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+                    self._stop_event.is_set()
+                    or time.time() - time_since_last_data > self.no_data_timeout
+                ):
+                    logger.debug(
+                        f"No more requests. Stopping. Sent {count} requests. {self.data_queue.qsize()} requests remaining."
+                    )
+                    await sift_stream.finish()
+                    return
+                else:
+                    time.sleep(self.IDLE_LOOP_PERIOD)
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:131]: asyncio.sleep so we only block this loop

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+
+        except asyncio.CancelledError:
+            # It's possible the thread was joined while sleeping waiting for data. Only note error if we have data left.
+            if self.data_queue.qsize() > 0:
+                logger.error(
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:136]: Should this be reraised so that it can be handled?

File: python/lib/sift_client/_internal/low_level_wrappers/rules.py [hunk 26]
```diff
@@ -0,0 +1,460 @@
...
+            rule_id: The rule ID to get.
+            client_key: The client key to get.
+
+        Returns:
+            The Rule.
+
+        Raises:
+            ValueError: If neither rule_id nor client_key is provided.
+        """
+        if rule_id is None and client_key is None:
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/rules.py:76]: Rely on the API for this

File: python/lib/sift_client/_internal/low_level_wrappers/rules.py [hunk 27]
```diff
@@ -0,0 +1,460 @@
...
+            update.channel_references
+            if "channel_references" in model_dump
+            else rule.channel_references
+        )
+        action = update.action if "action" in model_dump else rule.action
+        if bool(expression) != bool(channel_references):
+            raise ValueError(
+                "Expression and channel_references must both be provided or both be None"
+            )
+        assert channel_references is not None  # make mypy happy
```
Comment (alexluck-sift) [inline python/lib/sift_client/_internal/low_level_wrappers/rules.py:236]: should be able to put a type hint on `channel_references` when it is assigned above

Don't want to use an assertion here that if, for some reason, channel_references is none, would result in an assertion failure

File: python/lib/sift_client/types/ingestion.py [hunk 28]
```diff
@@ -0,0 +1,219 @@
...
+if TYPE_CHECKING:
+    from sift_client.client import SiftClient
+
+
+class IngestionConfig(BaseType[IngestionConfigProto, "IngestionConfig"]):
+    """
+    Model of the Sift Ingestion Config.
+    """
+
+    id: str
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/ingestion.py:37]: no longer needed

File: python/lib/sift_client/types/ingestion.py [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+    elif channel.data_type == ChannelDataType.UINT_32:
+        return IngestWithConfigDataChannelValuePy.uint32(value)
+    elif channel.data_type == ChannelDataType.UINT_64:
+        return IngestWithConfigDataChannelValuePy.uint64(value)
+    else:
+        raise ValueError(f"Invalid data type: {channel.data_type}")
+
+
+def _to_rust_type(data_type: ChannelDataType) -> ChannelDataTypePy:
+    # TODO: Make more elegant?
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/ingestion.py:191]: this can be part of the ChannelDataType class (and a switch statement is always a bit cleaner :)

File: python/lib/sift_client/types/rule.py (hunk 11 above)
Comment (alexluck-sift) [reply python/lib/sift_client/types/rule.py:36]: can we add some `raise NotImplementedError` stubs so we know to come back to these? or TODO statements?

File: python/lib/sift_client/types/rule.py [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+                if "ACTION_KIND_" + item.name == val:
+                    return item
+
+        return cls(int(val))
+
+
+class RuleAnnotationType(Enum):
+    """Enum for rule annotation types."""
+
+    UNSPECIFIED = AnnotationType.ANNOTATION_TYPE_UNSPECIFIED  # 0
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:174]: Not a huge fan of the enum to enum type, especially since the other type is the raw proto type.

I think it would make more sense to keep these as string enums and then having a mapping function like how we are handling other types

File: python/lib/sift_client/types/rule.py [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+    version_id: str | None = None
+    annotation_type: RuleAnnotationType | None = None
+    tags: List[str] | None = None
+    default_assignee_user_id: str | None = None
+
+    @classmethod
+    def annotation(
+        cls,
+        annotation_type: RuleAnnotationType,
+        tags: List[str] = [],
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:210]: mutable default argument shouldn't be used https://docs.python-guide.org/writing/gotchas/#what-you-should-do-instead

File: python/lib/sift_client/types/rule.py [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+                        annotation_type=self.annotation_type.value,  # type: ignore
+                    )
+                    if self.action_type == RuleActionType.ANNOTATION
+                    else None
+                ),
+            ),
+        )
+
+
+class RuleVersion(BaseModel):
```
Comment (alexluck-sift) [inline python/lib/sift_client/types/rule.py:272]: why are we using a different model than BaseType?

File: python/lib/sift_client/util/timestamp.py [hunk 33]
```diff
@@ -1,9 +1,17 @@
...
+from sift_stream_bindings import TimeValuePy
 
 
 def to_pb_timestamp(timestamp: datetime) -> Timestamp:
     timestamp_pb = Timestamp()
     timestamp_pb.FromDatetime(timestamp)
     return timestamp_pb
+
+
+def to_rust_py_timestamp(time: datetime) -> TimeValuePy:
```
Comment (alexluck-sift) [inline python/lib/sift_client/util/timestamp.py:13]: would be good to incorporate this into the sift_stream_bindings library directly

File: python/lib/sift_client/__init__.py [hunk 34]
```diff
@@ -1,7 +1,22 @@
...
+logger = logging.getLogger(__name__)
+logging.basicConfig(
+    level=logging.ERROR, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+)
+
+
+handler = logging.StreamHandler(sys.stdout)
+logger.addHandler(handler)
+
+# TODO: Remove
```
Comment (alexluck-sift) [inline python/lib/sift_client/__init__.py:21]: remove the logging here

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py (hunk 23 above)
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:98]: If it were set to 0, we'd immediately finish the stream before waiting for the first data point if one was not immediately present on starting the thread

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py (hunk 22 above)
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:86]: not an async function

**Review by alexluck-sift — COMMENTED** [review]
Resolved comments that have been addressed

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py (hunk 25 above)
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:136]: This is called during cleanup and don't wanna block other threads cleaning up

File: python/lib/sift_client/_internal/low_level_wrappers/ingestion.py (hunk 23 above)
Comment (ian-sift) [reply python/lib/sift_client/_internal/low_level_wrappers/ingestion.py:98]: can you be more specific?

File: python/lib/sift_client/types/rule.py (hunk 12 above)
Comment (ian-sift) [reply python/lib/sift_client/types/rule.py]: this is done

**Review by alexluck-sift — DISMISSED** [review]
Looks good to merge!

File: python/lib/sift_client/types/ingestion.py (hunk 29 above)
Comment (ian-sift) [reply python/lib/sift_client/types/ingestion.py:191]: switch (match) not till py3.10 so just gonna leave this as is for now. Also i thought you wanted all sift stream bindings stuff in the ingestion files not the channel types?

File: python/lib/sift_client/types/channel.py [hunk 35]
```diff
@@ -1,6 +1,461 @@
...
+
+
+# Channel config model
+class Channel(BaseType[ChannelProto, "Channel"]):
+    id: str | None = None
+    name: str
+    data_type: ChannelDataType
+    description: str | None = None
+    unit: str | None = None
+    component: str | None = None  # Deprecated
```
[the remark this answers is not in the document]
Comment (alexluck-sift) [reply python/lib/sift_client/types/channel.py]: this is handled on the backend

**Review by alexluck-sift — APPROVED** [review]
